monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/IntersperseObservable.scala

```diff
@@ -1,48 +1,74 @@
+/*
+ * Copyright (c) 2014-2016 by its authors. Some rights reserved.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package monix.reactive.internal.operators
 
 import monix.execution.Ack.Continue
-import monix.execution.atomic.AtomicBoolean
 import monix.execution.{Ack, Cancelable}
 import monix.reactive.Observable
 import monix.reactive.observers.Subscriber
 
 import scala.concurrent.Future
 
-/**
-  * Created by omainegra on 5/29/17.
-  */
-private[reactive] final class IntersperseObservable[+A](source: Observable[A],
-                                                        separator: A) extends Observable[A]{ self =>
+private[reactive] final class IntersperseObservable[+A](
+  source: Observable[A],
+  start: Option[A],
+  separator: A,
+  end: Option[A] ) extends Observable[A] { self =>
 
   override def unsafeSubscribeFn(out: Subscriber[A]): Cancelable = {
-    val firstTime = AtomicBoolean(true)
-    val isDone = AtomicBoolean(false)
-    var downstreamAck = Continue : Future[Ack]
-
     val upstream = source.unsafeSubscribeFn(new Subscriber[A] {
       implicit val scheduler = out.scheduler
 
+      private[this] var firstTime = true
+      private[this] var isDone = false
+      private[this] var downstreamAck = Continue : Future[Ack]
+
       override def onNext(elem: A): Future[Ack] = {
-        downstreamAck = if (firstTime.getAndSet(false)) {
-          out.onNext(elem)
+        downstreamAck = if (firstTime) {
+          firstTime = false
+          start.map(out.onNext).getOrElse(Continue).syncFlatMap {
+            case Continue => out.onNext(elem)
+            case ack => ack
+          }
         }
         else {
-          out.onNext(separator).flatMap {
+          out.onNext(separator).syncFlatMap {
             case Continue => out.onNext(elem)
             case ack => ack
           }
         }
+
         downstreamAck
       }
 
       def onError(ex: Throwable) = {
-        if (!isDone.getAndSet(true)){
+        if (!isDone){
+          isDone = true
           out.onError(ex)
         }
       }
       def onComplete() = {
-        downstreamAck.syncOnContinue {
-          if (!isDone.getAndSet(true)){
+        downstreamAck.syncFlatMap {
+          case Continue if end.isDefined => out.onNext(end.get)
+          case ack => ack
+        }.syncOnContinue {
```

Comment:
OK, I think I know what's going on. In `onNext` you can end up pushing 2 events instead of one, so if the subscriber is delaying on `sep`, you still have the `elem` queued up for sending it, so you can end up triggering `onNext(end)` and `onComplete` before that happens. See, I was wrong :-)

Btw, you need to check that you receive a `Continue` and don't need to do anything in case of `Stop` or `Failure(ex)`, so doing just one `syncOnContinue` there is totally fine.

```scala
ack.syncOnContinue {
  if (end.nonEmpty) out.onNext(end.get)
  out.onComplete
}
```
Or smth.